src/heartbeat.ts

```diff
@@ -0,0 +1,130 @@
+        this._register(vscode.tasks.onDidEndTask(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidEndTaskProcess(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeWorkspaceFolders(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeConfiguration(this.updateLastActivitiy, this));
+        this._register(vscode.languages.registerHoverProvider('*', {
+            provideHover: () => {
+                this.updateLastActivitiy();
+                return null;
+            }
+        }));
+
+        this.logger.trace(`Heartbeat manager for workspace ${workspaceId} (${instanceId}) - ${gitpodHost} started`);
+
+        // Start heatbeating interval
+        this.sendHeartBeat();
+        this.heartBeatHandle = setInterval(() => {
+            if (this.lastActivity + HeartbeatManager.HEARTBEAT_INTERVAL < new Date().getTime()) {
+                // no activity, no heartbeat
+                return;
+            }
+
+            this.sendHeartBeat();
+        }, HeartbeatManager.HEARTBEAT_INTERVAL);
+    }
+
+    private updateLastActivitiy() {
+        this.lastActivity = new Date().getTime();
+    }
+
+    private async sendHeartBeat(wasClosed?: true) {
+        const suffix = wasClosed ? 'closed heartbeat' : 'heartbeat';
+        try {
+            await withServerApi(this.accessToken, this.gitpodHost, async service => {
+                const workspaceInfo = await service.server.getWorkspace(this.workspaceId);
+                this.isWorkspaceRunning = workspaceInfo.latestInstance?.status?.phase === 'running';
```

Comment:
you should also check that instance id does not changed, as we discussed yesterday if a workspace is restarted then a user should go though browser again, we should not start sending heartbeat for restarted workspace